utils/land-patterns.stanza

```diff
@@ -239,52 +308,120 @@ public pcb-landpattern non-plated-hole-pkg (drill-r:Double, mask-r:Double) :
+  pads:Tuple<PinPadMapping>
+):
   inside pcb-landpattern:
-    
-    val n-col = to-int(ceil(to-double(n-pin) / to-double(n-row)))
-
-    val l0 = loc((to-double(n-row - 1) * row-pitch) / -2.0, (to-double(n-col - 1) * pin-pitch) / 2.0)
-    
-    for i in 0 to n-col do :
-      for j in 0 to n-row do :
-        pad p[n-row * i + j + 1] : pad-ref at l0 * loc(to-double(j) * row-pitch, (- to-double(i) * pin-pitch))
-    
-public defn make-pin-grid (n-pin:Int, n-row:Int, pin-pitch:Double, row-pitch:Double) :
-  make-pin-grid(n-pin, n-row, pin-pitch, row-pitch, pth-pad(0.5))
+    for p in pads do:
+      pad (pad-name(p)): (pad-shape(p) as Pad) at pose(p)
+
+; Create a generic landpattern with a uniform pad
+public defn make-generic-landpattern (
+  pad-names:Seqable<Ref>,
+  poses:Seqable<Pose>,
+  pad-shape:JITXDef
+):
+  val pads = 
+    for (pn in pad-names, pp in poses) seq:
+      PinPadMapping(pn, pad-shape, pp) 
+  make-generic-landpattern(to-tuple(pads))
+
+;===============================================================================
+;========================= Pad Arrangements ====================================
+;===============================================================================
+; Naming for a 2-column arrangment of pins 
+public defn soic-naming-convention (row:Int, column:Int, num-pins:Int) -> Ref:
+  fatal("soic naming convention supports a maximum of two columns") when (column >= 2 or column < 0)
+  val idx = row + 1 when column == 0 else num-pins - row
+  IndexRef(Ref("p"), idx)
+  
+; Create a two-column landpattern
+public defn make-two-column-landpattern (
+  count: Int,    ; number of pads in the column
+  pitch: Double, ; pitch of the column
+  naming-convention: (Int, Int) -> Ref, ; naming convention
+  pad-convention: (Int, Int) -> JITXDef,
+  spacing:Double,
+):
+  val mapping = 
+    for col in 0 to 2 seq-cat:
+      val column = pose-column(loc(to-double(col) * spacing, 0.0), count, pitch)
+      for (pose in column, row in 0 to false) seq:
+        PinPadMapping(naming-convention(row, col), pad-convention(row, col), pose)
+  make-generic-landpattern(to-tuple(mapping))
+
+public defn make-grid-landpattern (
+  num-rows:Int,
+  num-cols:Int,
+  row-pitch:Double,
+  col-pitch:Double,
+  naming-convention: (Int, Int) -> Ref,
+  pad-convention: (Int, Int) -> JITXDef,
+):
+  make-grid-landpattern(num-rows, num-cols, row-pitch, col-pitch, naming-convention, pad-convention, fn (r, c): true)
+
+; Create a grid landpattern
+public defn make-grid-landpattern (
+  num-rows:Int,
+  num-cols:Int,
+  row-pitch:Double,
+  col-pitch:Double,
+  naming-convention: (Int, Int) -> Ref,
+  pad-convention: (Int, Int) -> JITXDef,
+  ignore?: (Int, Int) -> True|False,
+):
+  make-grid-landpattern(loc(0.0, 0.0), num-rows, num-cols, row-pitch, col-pitch, naming-convention, pad-convention, ignore?)
+
+; Create a grid pattern with arbitrary origin
+public defn make-grid-landpattern (
+  origin: Pose,
```

Comment:
the "origin" is the position of pin 0, in the top left corner of the landpattern.